### ISSUE TITLE
[numpy] Change the existing infer shape and utility functions to use the new definition of unknown shapes

### ISSUE THREAD

**Issue by reminisce** [apache/incubator-mxnet#14323]
### Problem
**Current defition of unknown shapes**
`shape.ndim() = 0` or any `shape[i] = 0`, for `0 <= i < shape.ndim()`.

should be changed to

**new definition of unknown shapes**
`shape.ndim() = -1` or any `shape[i] = -1`, for `0 <= i < shape.ndim()`.

**How to make changes?**
Always use util function `mxnet::op::shape_is_known(const mxnet::TShape&)` from https://github.com/apache/incubator-mxnet/pull/14315 to check whether the shape info is complete.

### **File list**
Listed header files only. Note that some operators register infer shape functions in `.cc` files, should use the header files as reference points to change the corresponding infer shape functions defined in `.cc`.

**src/operator** @reminisce 
- [ ] batch_norm_v1-inl.h
- [ ] bilinear_sampler-inl.h
- [ ] c_lapack_api.h
- [ ] channel_op_common.h
- [ ] convolution_v1-inl.h
- [ ] correlation-inl.h
- [ ] crop-inl.h
- [ ] cudnn_bilinear_sampler-inl.h
- [ ] cudnn_lrn-inl.h
- [ ] cudnn_rnn-inl.h
- [ ] cudnn_spatial_transformer-inl.h
- [ ] elemwise_op_common.h
- [ ] grid_generator-inl.h
- [ ] identity_attach_KL_sparse_reg-inl.h
- [ ] instance_norm-inl.h
- [ ] l2_normalization-inl.h
- [ ] leaky_relu-inl.h
- [ ] linalg.h
- [ ] linalg_impl.h
- [ ] loss_binary_op-inl.h
- [ ] make_loss-inl.h
- [ ] math_functions-inl.h
- [ ] mshadow_op.h
- [ ] mxnet_op.h
- [ ] operator_common.h
- [ ] operator_tune-inl.h
- [ ] operator_tune.h
- [ ] optimizer_op-inl.h
- [ ] pad-inl.h
- [ ] pooling_v1-inl.h
- [ ] regression_output-inl.h
- [ ] rnn-inl.h
- [ ] rnn_impl.h
- [ ] roi_pooling-inl.h
- [ ] sequence_last-inl.h
- [ ] sequence_mask-inl.h
- [ ] sequence_op_common.h
- [ ] sequence_reverse-inl.h
- [ ] slice_channel-inl.h
- [ ] softmax_output-inl.h
- [ ] spatial_transformer-inl.h
- [ ] special_functions-inl.h
- [ ] subgraph_op_common.h
- [ ] svm_output-inl.h
- [ ] swapaxis-inl.h

**src/operator/contrib** @junrushao1994 
- [ ] adamw-inl.h
- [ ] adaptive_avg_pooling-inl.h
- [ ] bilinear_resize-inl.h
- [ ] boolean_mask-inl.h
- [ ] bounding_box-inl.h
- [ ] count_sketch-inl.h
- [ ] deformable_convolution-inl.h
- [ ] deformable_psroi_pooling-inl.h
- [ ] dgl_graph-inl.h
- [ ] erfinv-inl.h
- [ ] fft-inl.h
- [ ] ifft-inl.h
- [ ] index_copy-inl.h
- [ ] krprod.h
- [ ] multi_proposal-inl.h
- [ ] multibox_detection-inl.h
- [ ] multibox_prior-inl.h
- [ ] multibox_target-inl.h
- [ ] nnvm_to_onnx-inl.h
- [ ] optimizer_op-inl.h
- [ ] proposal-inl.h
- [ ] psroi_pooling-inl.h
- [ ] quadratic_op-inl.h
- [ ] roi_align-inl.h
- [ ] sync_batch_norm-inl.h
- [ ] tensorrt-inl.h
- [ ] transformer-inl.h

**src/operator/custom** @junrushao1994 

- [ ] custom-inl.h
- [ ] native_op-inl.h
- [ ] ndarray_op-inl.h

**src/operator/image** @reminisce  

- [ ] image_random-inl.h
- [ ] image_utils.h
- [ ] resize-inl.h

**src/operator/nn** @reminisce  

- [ ] activation-inl.h
- [ ] batch_norm-inl.h
- [ ] concat-inl.h
- [ ] convolution-inl.h
- [ ] ctc_loss-inl.h
- [ ] deconvolution-inl.h
- [ ] depthwise_convolution-inl.h
- [ ] dropout-inl.h
- [ ] fully_connected-inl.h
- [ ] im2col.h
- [ ] layer_norm-inl.h
- [ ] lrn-inl.h
- [ ] pool.h
- [ ] pool_utils.h
- [ ] pooling-inl.h
- [ ] sequence_mask-inl.h
- [ ] softmax-inl.h
- [ ] softmax_activation-inl.h
- [ ] upsampling-inl.h

**src/operator/quantization** @reminisce  
- [ ] dequantize-inl.h
- [ ] quantization_utils.h
- [ ] quantize-inl.h
- [ ] quantize_v2-inl.h
- [ ] quantized_flatten-inl.h
- [ ] requantize-inl.h

**src/operator/quantization/mkldnn** @reminisce 

- [ ] mkldnn_dequantize-inl.h
- [ ] mkldnn_quantize-inl.h
- [ ] mkldnn_quantize_v2-inl.h
- [ ] mkldnn_requantize-inl.h

**src/operator/random** @reminisce 

- [ ] multisample_op.h
- [ ] sample_multinomial_op.h
- [ ] sample_op.h
- [ ] sampler.h
- [ ] unique_sample_op.h

**src/operator/subgraph/mkldnn** @reminisce 

- [ ] mkldnn_conv-inl.h

**src/operator/tensor** @reminisce 
- [ ] broadcast_reduce-inl.h
- [ ] broadcast_reduce_op.h
- [ ] cast_storage-inl.h
- [ ] control_flow_op.h
- [ ] diag_op-inl.h
- [ ] dot-inl.h
- [ ] elemwise_binary_broadcast_op.h
- [ ] elemwise_binary_op-inl.h
- [ ] elemwise_binary_op.h
- [ ] elemwise_binary_scalar_op.h
- [ ] elemwise_scatter_op.h
- [ ] elemwise_sum.h
- [ ] elemwise_unary_op.h

**src/operator/tensor** @reminisce  
- [ ] histogram-inl.h
- [ ] indexing_op.h
- [ ] init_op.h
- [ ] la_op-inl.h
- [ ] la_op.h
- [ ] matrix_op-inl.h
- [ ] ordering_op-inl.h
- [ ] ravel.h
- [ ] slice-inl.h
- [ ] sort_op.h
- [ ] sparse_retain-inl.h
- [ ] square_sum-inl.h

**Comment by piyushghai** [comment]
@mxnet-label-bot Add [Numpy, Good first issue]

**Comment by junrushao1994** [comment]
@reminisce Will we PR to the numpy branch?

**Comment by reminisce** [comment]
@junrushao1994 Yes.

**Comment by junrushao1994** [comment]
#14409
